src/controlplane/scraper.go

```diff
@@ -0,0 +1,270 @@
+)
+
+// Providers is a struct holding pointers to all the clients Scraper needs to get data from.
+// TODO: Extract this out of the package.
+type Providers struct {
+	K8s kubernetes.Interface
+}
+
+// Scraper takes care of getting metrics from an autodiscovered CP instances.
+type Scraper struct {
+	Providers
+	logger               log.Logger
+	config               *config.Config
+	k8sVersion           *version.Info
+	components           []component
+	informerClosers      []chan<- struct{}
+	podListerByNamespace map[string]v1.PodLister
+	inClusterConfig      *rest.Config
+}
+
+// ScraperOpt are options that can be used to configure the Scraper
+type ScraperOpt func(s *Scraper) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) ScraperOpt {
+	return func(s *Scraper) error {
+		if logger == nil {
+			return fmt.Errorf("logger canont be nil")
+		}
+		s.logger = logger
+		return nil
+	}
+}
+
+// WithRestConfig returns an OptionFunc to change the restConfig from default empty config.
+func WithRestConfig(restConfig *rest.Config) ScraperOpt {
+	return func(s *Scraper) error {
+		if restConfig == nil {
+			return fmt.Errorf("restConfig canont be nil")
+		}
+		s.inClusterConfig = restConfig
+		return nil
+	}
+}
+
+// Close will signal internal informers to stop running.
+func (s *Scraper) Close() {
+	for _, ch := range s.informerClosers {
+		close(ch)
+	}
+}
+
+// NewScraper builds a new Scraper, initializing its internal informers. After use, informers should be closed by calling Close().
+func NewScraper(config *config.Config, providers Providers, options ...ScraperOpt) (*Scraper, error) {
+	var err error
+	s := &Scraper{
+		config:    config,
+		Providers: providers,
+		// TODO: An empty implementation of the logger interface would be better
+		logger:               log.New(false, io.Discard),
+		podListerByNamespace: make(map[string]v1.PodLister),
+		components:           newComponents(config.ControlPlane),
+		inClusterConfig:      &rest.Config{},
+	}
+
+	for i, opt := range options {
+		if err := opt(s); err != nil {
+			return nil, fmt.Errorf("applying config option #%d: %w", i, err)
+		}
+	}
+
+	// TODO If this could change without a restart of the pod we should run it each time we scrape data,
+	// possibly with a reasonable cache Es: NewCachedDiscoveryClientForConfig
+	s.k8sVersion, err = providers.K8s.Discovery().ServerVersion()
+	if err != nil {
+		return nil, fmt.Errorf("fetching K8s version: %w", err)
+	}
+
+	s.buildLister()
+
+	return s, nil
+}
+
+// Run scraper collect the data populating the integration entities
+func (s *Scraper) Run(i *integration.Integration) error {
+	var jobs []*scrape.Job
+	for _, component := range s.components {
+
+		job, err := s.externalEndpoint(component)
+		if err != nil {
+			return fmt.Errorf("configuring %q external endpoint: %w", component.Name, err)
+		}
+
+		if component.StaticEndpointConfig == nil {
```

Comment:
Shouldn't this be put first on the loop and then on the externalEndpoint() removed the first condition? I point it bellow.